src/bindfs.c

```diff
@@ -1041,47 +1038,52 @@ static int bindfs_access(const char *path, int wants)
  *
  * @param path The path to iterate over.
  *
- * @return 0 if all path components have the search permission, -EACCES
+ * @return true if all path components have the search permission, false
  *     otherwise.
  */
-static int check_path_search_perms(const char *path) {
-    size_t path_length = strlen(path);
-    char *part = malloc(path_length + 1);
-    memcpy(part, path, path_length + 1);
+static bool path_has_search_perms(const char *path) {
+    char *part = strdup(path);
 
-    int access_check_result;
+    bool parent_dir = true;
     do {
         part = dirname(part);
+
+        if (parent_dir && access_check(part, W_OK) != 0) {
+            free(part);
+            return false;
+        }
+
         DPRINTF(
             "Checking path component for required search permissions: %s",
             part);
 
-        access_check_result = access_check(part, X_OK);
-        if (access_check_result != 0) {
+        if (access_check(part, X_OK) != 0) {
             DPRINTF(
                 "Path component doesn't have required search permission: %s",
                 part);
-            return access_check_result;
+            free(part);
+            return false;
         }
+
+        parent_dir = false;
     } while(strcmp(part, "/") != 0);
 
     DPRINTF("Path has search required permissions: %s", path);
-    return 0;
+    free(part);
```

Comment:
After `path = dirname(path)`, `path` is no longer the thing you allocated so `free`'ing it is wrong. You need to keep the original `strdup`'ed pointer around.